services/editor/src/pages/context/components/FixedKeys/FixedKeys.js

```diff
@@ -1,138 +1,85 @@
-import React, { Component } from 'react';
-import PropTypes from 'prop-types';
+import React from 'react';
+import { connect } from 'react-redux';
+import { compose, mapProps } from 'recompose';
 import R from 'ramda';
 import classnames from 'classnames';
+import deepEqual from 'deep-equal';
+import * as contextActions from '../../../../store/ducks/context';
+import { getFixedKeys, FIXED_PREFIX } from '../../../../services/context-service';
 import SaveButton from '../../../../components/common/SaveButton/SaveButton';
 import FixedKeysList from './FixedKeysList/FixedKeysList';
 import './FixedKeys.css';
 
-function calculateKeys(fixedKeys) {
-  const result = Object.entries(fixedKeys).map(([key, value]) => ({
-    remote: { key, value },
-    local: { key, value },
-    isRemoved: false,
-  }));
-
-  if (result.length === 0) {
-    return [
-      {
-        local: { key: '', value: '' },
-        isRemoved: false,
-      },
-    ];
-  }
-
-  return result;
+const mapWithProp = prop =>
+  R.pipe(Object.entries, R.map(([keyPath, value]) => ({ keyPath, [prop]: value })));
+
+function extractKeys(remote, local) {
+  return R.pipe(
+    R.mapObjIndexed((items, prop) => mapWithProp(prop)(items)),
+    R.values,
+    R.flatten,
+    R.groupBy(R.prop('keyPath')),
+    R.map(R.mergeAll),
+    R.values,
+  )({ remote, local });
 }
 
-function exists(obj) {
-  return obj !== undefined && obj.toString() !== '';
-}
-
-class FixedKeys extends Component {
-  constructor(props) {
-    super(props);
-
-    const keys = calculateKeys(props.fixedKeys);
-    this.state = {
-      keys,
-      hasChanges: false,
+const getProperties = R.pickBy((_, prop) => !prop.startsWith(FIXED_PREFIX));
+
+const extractLocal = R.pipe(
+  R.filter(({ local }) => local !== undefined),
+  R.indexBy(({ keyPath }) => FIXED_PREFIX + keyPath),
+  R.map(R.prop('local')),
+);
+
+const hasValues = R.pipe(Object.entries, R.all(R.all(x => x !== undefined && x !== '')));
```

Comment:
it's an array of arrays, I can do `R.flatten` and then only one `R.all` if it's more undertandable